File: ports/csharp/src/network/kernel/BcsNetwork.cs
using System;
using Bcs.Core.Kernel;

namespace Bcs.Network.Kernel
{
    public class BcsHostAddress : BcsObject
    {
        private string _address;
        private bool _isIPv6;
        private readonly object _lock = new object();

        public BcsHostAddress(string address) : base(null)
        {
            _address = address;
        }

        public override string ToString()
        {
            lock (_lock) { return _address; }
        }
    }

    public enum ProxyType
    {
        DefaultProxy = 0,
        Socks5Proxy,
        NoProxy,
        HttpProxy,
        HttpCachingProxy,
        FtpCachingProxy
    }

    public class BcsNetworkProxy : BcsObject
    {
        private ProxyType _proxyType = ProxyType.DefaultProxy;
        private string _host = string.Empty;
        private int _port = 0;
        private readonly object _lock = new object();

        public BcsNetworkProxy() : base(null) {}

        public ProxyType Type
        {
            get { lock (_lock) { return _proxyType; } }
            set { lock (_lock) { _proxyType = value; } }
        }

        public string HostName
        {
            get { lock (_lock) { return _host; } }
            set { lock (_lock) { _host = value; } }
        }

        public int Port
        {
            get { lock (_lock) { return _port; } }
            set { lock (_lock) { _port = value; } }
        }
    }

    public class BcsNetworkManager : BcsObject
    {
        private static readonly Lazy<BcsNetworkManager> _instance = new Lazy<BcsNetworkManager>(() => new BcsNetworkManager());
        public static BcsNetworkManager Instance => _instance.Value;

        private BcsNetworkProxy _applicationProxy;
        private readonly object _lock = new object();

        private BcsNetworkManager() : base(null) {}

        public BcsNetworkProxy ApplicationProxy
        {
            get { lock (_lock) { return _applicationProxy; } }
            set { lock (_lock) { _applicationProxy = value; } }
        }
    }
}

File: ports/rust/src/network/kernel/bcs_network.rs
use std::sync::{Arc, Mutex, OnceLock};
use crate::core::kernel::bcs_object::BcsObject;

pub struct BcsHostAddress {
    pub base: Arc<BcsObject>,
    address: Mutex<String>,
    is_ipv6: Mutex<bool>,
}

impl BcsHostAddress {
    pub fn new(address: &str) -> Arc<Self> {
        Arc::new(Self {
            base: BcsObject::new(),
            address: Mutex::new(address.to_string()),
            is_ipv6: Mutex::new(false),
        })
    }

    pub fn to_string(&self) -> String {
        self.address.lock().unwrap().clone()
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ProxyType {
    DefaultProxy = 0,
    Socks5Proxy,
    NoProxy,
    HttpProxy,
    HttpCachingProxy,
    FtpCachingProxy,
}

pub struct BcsNetworkProxy {
    pub base: Arc<BcsObject>,
    proxy_type: Mutex<ProxyType>,
    host: Mutex<String>,
    port: Mutex<u16>,
}

impl BcsNetworkProxy {
    pub fn new() -> Arc<Self> {
        Arc::new(Self {
            base: BcsObject::new(),
            proxy_type: Mutex::new(ProxyType::DefaultProxy),
            host: Mutex::new(String::new()),
            port: Mutex::new(0),
        })
    }

    pub fn set_type(&self, t: ProxyType) {
        *self.proxy_type.lock().unwrap() = t;
    }

    pub fn set_host_name(&self, host: &str) {
        *self.host.lock().unwrap() = host.to_string();
    }

    pub fn set_port(&self, port: u16) {
        *self.port.lock().unwrap() = port;
    }
}

pub struct BcsNetworkManager {
    pub base: Arc<BcsObject>,
    application_proxy: Mutex<Option<Arc<BcsNetworkProxy>>>,
}

static INSTANCE: OnceLock<Arc<BcsNetworkManager>> = OnceLock::new();

impl BcsNetworkManager {
    pub fn instance() -> Arc<BcsNetworkManager> {
        INSTANCE.get_or_init(|| {
            Arc::new(Self {
                base: BcsObject::new(),
                application_proxy: Mutex::new(None),
            })
        }).clone()
    }

    pub fn set_application_proxy(&self, proxy: Arc<BcsNetworkProxy>) {
        *self.application_proxy.lock().unwrap() = Some(proxy);
    }
}
